File: src/main/java/com/portofino/realtrainmodunofficial/network/SpeakerPlayPayload.java
package com.portofino.realtrainmodunofficial.network;

import com.portofino.realtrainmodunofficial.RealTrainModUnofficial;
import com.portofino.realtrainmodunofficial.client.sound.LegacyScriptSoundManager;
import io.netty.buffer.ByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.neoforged.neoforge.network.handling.IPayloadContext;

/**
 * スピーカーがレッドストーン信号を受けて鳴るとき、サーバーが範囲内プレイヤーへ送る再生パケット。
 * クライアントは指定座標で {@code soundId} を再生する。
 * volume は可聴範囲(ブロック)から算出済みの値（MC の LINEAR 減衰は概ね volume×16 ブロック）。
 */
public record SpeakerPlayPayload(double x, double y, double z, String soundId, float volume, float pitch)
        implements CustomPacketPayload {

    public static final Type<SpeakerPlayPayload> TYPE = new Type<>(
        ResourceLocation.fromNamespaceAndPath(RealTrainModUnofficial.MODID, "speaker_play")
    );

    public static final StreamCodec<ByteBuf, SpeakerPlayPayload> STREAM_CODEC = StreamCodec.composite(
        ByteBufCodecs.DOUBLE,
        SpeakerPlayPayload::x,
        ByteBufCodecs.DOUBLE,
        SpeakerPlayPayload::y,
        ByteBufCodecs.DOUBLE,
        SpeakerPlayPayload::z,
        ByteBufCodecs.STRING_UTF8,
        SpeakerPlayPayload::soundId,
        ByteBufCodecs.FLOAT,
        SpeakerPlayPayload::volume,
        ByteBufCodecs.FLOAT,
        SpeakerPlayPayload::pitch,
        SpeakerPlayPayload::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public static void handleOnClient(SpeakerPlayPayload payload, IPayloadContext context) {
        context.enqueueWork(() ->
            LegacyScriptSoundManager.playAt(
                payload.x(), payload.y(), payload.z(),
                payload.soundId(), payload.volume(), payload.pitch()
            )
        );
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/network/SpeakerPlayPayload.kt
package cc.mirukuneko.realtrainmodrenewed.network

import cc.mirukuneko.realtrainmodrenewed.RealTrainModRenewed
import cc.mirukuneko.realtrainmodrenewed.client.sound.LegacyScriptSoundManager
import io.netty.buffer.ByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import net.neoforged.neoforge.network.handling.IPayloadContext

/** Server-to-client packet that plays a speaker sound at a fixed position. */
@JvmRecord
data class SpeakerPlayPayload(
    val x: Double,
    val y: Double,
    val z: Double,
    val soundId: String,
    val volume: Float,
    val pitch: Float,
) : CustomPacketPayload {
    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    companion object {
        @JvmField
        val TYPE: CustomPacketPayload.Type<SpeakerPlayPayload> = CustomPacketPayload.Type(
            Identifier.fromNamespaceAndPath(RealTrainModRenewed.MODID, "speaker_play")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<ByteBuf, SpeakerPlayPayload> = StreamCodec.composite(
            ByteBufCodecs.DOUBLE,
            { payload -> payload.x },
            ByteBufCodecs.DOUBLE,
            { payload -> payload.y },
            ByteBufCodecs.DOUBLE,
            { payload -> payload.z },
            ByteBufCodecs.STRING_UTF8,
            { payload -> payload.soundId },
            ByteBufCodecs.FLOAT,
            { payload -> payload.volume },
            ByteBufCodecs.FLOAT,
            { payload -> payload.pitch },
            ::SpeakerPlayPayload,
        )

        @JvmStatic
        fun handleOnClient(payload: SpeakerPlayPayload, context: IPayloadContext) {
            context.enqueueWork {
                LegacyScriptSoundManager.playAt(
                    payload.x,
                    payload.y,
                    payload.z,
                    payload.soundId,
                    payload.volume,
                    payload.pitch,
                )
            }
        }
    }
}
